File: 648-replace-words/main.go
package main

func replaceWords(dictionary []string, sentence string) string {

    rootSet := make(map[string]bool)
    for _, word := range dictionary {
        rootSet[word] = true
    }

    sentenceList := strings.Split(sentence, " ")

    var ans []string
    for _, word := range sentenceList {
        ans = append(ans, findRoot(word, rootSet))
    }

    return strings.Join(ans, " ")

}

func findRoot(word string, rootSet map[string]bool) string {

    for i := 1; i < len(word) + 1; i++ {
        prefix := word[:i]
        if _, ok := rootSet[prefix]; ok {
            return prefix
        }
    }
    return word
}

File: 648-replace-words/main.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:

        root_set = set(dictionary)
        
        def findRoot(word: str) -> str:
            for i in range(1, len(word) + 1):
                prefix = word[:i]
                if prefix in root_set:
                    return prefix
            return word
        
        words = sentence.split()
        replaced_words = map(findRoot, words)

        return ' '.join(replaced_words)
